Create a report that flags missing and modified rules

gregvanbrug opened the pull request:
This update prints a report that allows you to see which rules have not been defined in our `eslint` config and lets us know which ones we've modified.

Tests are added to ensure the reporter is "reporting" correctly.

If accepted, the next step would be to ensure all the missing rules are defined in our config so that we no longer need to extend the standard.

Fixes: https://apostrophecms.atlassian.net/jira/software/projects/ABP/boards/8?selectedIssue=ABP-5

boutell commented:
after we change to not extending the standard, how will we know if the standard is improved in notable ways, such as the availability of checks for things that previously could not be linted?

gregvanbrug commented:
Ah. I thought just flagging the differences would do it, but you're right that might not be super clear as to what's new vs. what's different. 

Maybe instead of installing latest, we could do a minimum version and check to see if the installed one is greater than the one we've specified. Could use that to alert the user there's a new version and maybe even show part of the changelog.

Maybe running that on postinstall so it's noisy enough to be seen but not too noisy to be annoying.

abea commented on package.json:
```diff
@@ -15,5 +18,18 @@
+  "peerDependencies": {
+    "eslint": "^7.1.0"
+  },
+  "dependencies": {
```
How does this jive with the current readme?

> Due to limitations of eslint the dependencies of eslint plugins like this one
must be explicitly loaded in your `package.json`.
> Here is the necessary section of `package.json`:
```javascript
"devDependencies": {
    "mocha": "^5.1.1",
    "eslint": "^4.0.0",
    "eslint-config-standard": "^11.0.0",
    "eslint-plugin-import": "^2.13.0",
    "eslint-plugin-node": "^6.0.1",
    "eslint-plugin-promise": "^3.8.0",
    "eslint-plugin-standard": "^3.1.0"
  }
```

gregvanbrug replied:
The readme probably needs to be updated. ESLint recommends doing it that way: https://eslint.org/docs/developer-guide/shareable-configs#publishing-a-shareable-config

I'm not 100% sure but also don't think they need to include `mocha` in their local package.json since that's just for testing. Semi wondering if this could also be automated...

boutell commented:
I thought not needing to extend anything was the ultimate goal.

On Wed, Jun 3, 2020 at 3:54 PM Greg van Brug <notifications@github.com>
wrote:

> *@gregvanbrug* commented on this pull request.
> ------------------------------
>
> In package.json
> <https://github.com/apostrophecms/eslint-config-apostrophe/pull/3#discussion_r434817029>
> :
>
> > @@ -15,5 +18,18 @@
>    "bugs": {
>      "url": "https://github.com/apostrophecms/eslint-config-apostrophe/issues"
>    },
> -  "homepage": "https://github.com/apostrophecms/eslint-config-apostrophe#readme"
> +  "homepage": "https://github.com/apostrophecms/eslint-config-apostrophe#readme",
> +  "peerDependencies": {
> +    "eslint": "^7.1.0"
> +  },
> +  "dependencies": {
>
> The readme probably needs to be updated. ESLint recommends doing it that
> way:
> https://eslint.org/docs/developer-guide/shareable-configs#publishing-a-shareable-config
>
> I'm not 100% sure but also don't think they need to include mocha in
> their local package.json since that's just for testing. Semi wondering if
> this could also be automated...
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/apostrophecms/eslint-config-apostrophe/pull/3#discussion_r434817029>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAH27IDZ5GWR7IGBNQ5TF3RU2S5RANCNFSM4NRC3OGA>
> .
>


-- 

THOMAS BOUTELL | CHIEF TECHNOLOGY OFFICER
APOSTROPHECMS | apostrophecms.com | he/him/his

gregvanbrug commented:
@boutell @abea Yes. Once we have all the missing rules added to our config we basically have a fork of the standard rules so we wouldn't need to extend the the standard anymore. 

I added a "check" task yesterday that will let us know about updates to ESLint standard. It can be run manually but it will also run on `postinstall` so we should know if we're up to date or need to add something. Between that check and the missing and modified reports we should know exactly how we are different than the standard and what new things we should add a rule for.

abea commented on lib/checkUpdates.js:
```diff
@@ -0,0 +1,40 @@
+    Updates available for the following configs: \n
+${JSON.stringify(availableUpdates, null, 4)}\n`);
+  } else {
+    console.log("    🎉 ESLint config is up to date.");
```
I always love to see emojis in log messages. 🤗

abea commented:
@gregvanbrug I cleaned up the merge conflict and added a changelog entry, but I'll leave it to you to merge in case you want to adjust either.

gregvanbrug commented:
lgtm, merging.
